proposals/nullable-reference-types-specification.md

```diff
@@ -0,0 +1,266 @@
+
+- An unconstrained type parameter `T` is *unknown*
+- A type parameter `T` where all constraints are *nullable*, *unknown* or the `class?` constraint is *unknown*
+- A type parameter `T` where at least one constraint is *oblivious* or *nonnullable* or one of the `struct` or `class` constraints is
+    - *oblivious* in a *disabled* annotation context
+    - *nonnullable* in an *enabled* annotation context
+- A nullable type parameter `T?` where at least one of `T`'s constraints is *oblivious* or *nonnullable* or one of the `struct` or `class` constraints, is
+    - *nullable* in a *disabled* annotation context (but a warning is yielded)
+    - *nullable* in an *enabled* annotation context
+
+For a type parameter `T`, `T?` is only allowed if `T` is known to be a value type or known to be a reference type.
+
+
+# Constraints
+
+Nullable reference types can be used as generic constraints. Furthermore `object` is now valid as an explicit constraint. Absence of a constraint is now equivalent to an `object?` constraint (instead of `object`), but (unlike `object` before) `object?` is not prohibited as an explicit constraint.
+
+`class?` is a new constraint denoting "possibly nullable reference type", whereas `class` denotes "nonnullable reference type".
+
+The nullability of a type argument or of a constraint does not impact whether the type satisfies the constraint, except where that is already the case today (nullable value types do not satisfy the `struct` constraint). However, if the type argument does not satisfy the nullability requirements of the constraint, a warning may be given.
+
+
+# Null state and null tracking
+
+Every expression in a given source location has a *null state*, which indicated whether it is believed to potentially evaluate to null. The null state is either "not null" or "maybe null". The null state is used to determine whether a warning should be given about null-unsafe conversions and dereferences.
+
+## Null tracking for variables
+
+For certain expressions denoting variables or properties, the null state is tracked between occurrences, based on assignments to them, tests performed on them and the control flow between them. This is similar to how definite assignment is tracked for variables. The tracked expressions are the ones of the following form:
+
+```antlr
+tracked_expression
+    : simple_name
+    | this
+    | base
+    | tracked_expression '.' identifier
+    ;
+```
+
+Where the identifiers denote fields or properties.
+
+***Describe null state transitions similar to definite assignment***
+
+## Null state for expressions
+
+The null state of an expression is derived from its form and type, and from the null state of variables involved in it.
+
+## Literals
+
+The null state of a `null` literal is "maybe null". The null state of a `default` literal that is being converted to a type that is known not to be a nonnullable value type is "maybe null". The null state of any other literal is "not null".
+
```

Comment:
as per above, this is where 'definitely null' might be a useful 'null state'.  There will likely be times something might-be null, and i'm ok with a warning, but would like to update to 'error' if 'definitely null'.